profiling/src/profile/mod.rs

```diff
@@ -525,21 +543,41 @@ impl Profile {
             .endpoints
             .mappings
             .get(&local_root_span_id)
-            .map(StringId::clone))
+            .map(|v| Label::str(self.endpoints.endpoint_label, *v)))
+    }
+
+    fn get_endpoint_for_labels(&self, label_set_id: LabelSetId) -> anyhow::Result<Option<Label>> {
+        let label = self.get_label_set(label_set_id).iter().find_map(|id| {
+            let label = self.get_label(*id);
+            if label.get_key() == self.endpoints.local_root_span_id_label {
+                Some(label)
+            } else {
+                None
+            }
+        });
+        if let Some(label) = label {
+            self.get_endpoint_for_label(label)
+        } else {
+            Ok(None)
+        }
     }
 
     fn translate_and_enrich_sample_labels(
         &self,
-        sample: &Sample,
+        sample: Sample,
+        timestamp: Option<Timestamp>,
     ) -> anyhow::Result<Vec<pprof::Label>> {
-        let mut labels: Vec<_> = sample.labels.iter().map(pprof::Label::from).collect();
-        if let Some(offset) = sample.local_root_span_id_label_offset {
-            // Safety: this offset was created internally and isn't be mutated.
-            let lrsi_label = unsafe { sample.labels.get_unchecked(offset) };
-            if let Some(endpoint_value_id) = self.get_endpoint_for_label(lrsi_label)? {
-                labels.push(Label::str(self.endpoints.endpoint_label, endpoint_value_id).into());
-            }
-        }
+        let labels: Vec<_> = self
+            .get_label_set(sample.labels)
+            .iter()
+            .map(|l| self.get_label(*l).into())
+            .chain(
+                self.get_endpoint_for_labels(sample.labels)?
+                    .map(pprof::Label::from),
+            )
```

Comment:
Since the list of labels is quite small, maybe this is not a big difference, but it seems kinda weird that we're iterating through the labels twice: once to covert them into `pprof::Label`s, and then again to search for the `local root span id` label; in the old code we actually did both intertwined -- conversion and finding the label.